### ISSUE TITLE
About torch.matmul

### ISSUE THREAD

**Issue by Xin-yang-Liu** [pytorch/pytorch#60234]
## 🐛 Bug

I tried to utilize the boardcast feature in torch.matmul, but here are some numerical errors when involves certain types of sparse matrix. To be clear, I did not use any sparse matrix format but most elements in the matrix ```lapa_c``` are zero (see the sample code I provide below).  

## To Reproduce

```Python
import torch

def Build_CoefficientMatrix(c:list,meshx:int,periodic:bool=False):
    '''
    c is list of FD coefficient
    e.g. for 1st derivative with 2nd accuracy central difference:
    c=[-0.5,0] 
    '''
    if 2*len(c)-1>=meshx: raise ValueError
    acc = len(c)   
    
    tmp=[]
    c.reverse()
    for i in range(acc):
        x = torch.cat((torch.cat((torch.zeros((i,meshx-i)),
                                    c[i]*torch.eye(meshx-i)),dim=0),
                                    torch.zeros((meshx,i))
                                    ),dim=1)
        tmp.append(x)
    re=tmp[0]
    for k in tmp[1:]:
        re+=k+k.T

    if periodic:
        re[:acc,-acc:]=re[acc:2*acc,:acc]
        re[-acc:,:acc]=re[:acc,acc:2*acc]
    return re

lapa_c=Build_CoefficientMatrix([1/90,-3/20,3/2,-49/18],64,periodic=True)

x=torch.rand(2,64,1)
out1=torch.matmul(lapa_c,x)
out2=torch.matmul(lapa_c,x[0,:,0])
print(out1[0,0,0].item())
print(out2[0].item())
```
The ```Build_CoefficientMatrix``` is used to build a [band matrix](https://en.wikipedia.org/wiki/Band_matrix) (with some non-zero elements in the bottom left and upper right corner when ```periodic=True```)

## Expected behavior
The two printed outputs are supposed to be the same, but actually they are not. 

##
if you change: 
* the batch size of ```x``` back to ```1```(```x=torch.rand(1,64,1)```)
* the ```periodic``` parameter in ```Build_CoefficientMatrix``` to ```False```
* the ```c``` parameter in ```Build_CoefficientMatrix``` to a list with length no greater than 3

the two results match each other.

## Environment

PyTorch version: 1.8.1
Is debug build: False
CUDA used to build PyTorch: 11.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.2 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.8 (64-bit runtime)
Python platform: Linux-5.8.0-55-generic-x86_64-with-glibc2.10
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce RTX 3070
Nvidia driver version: 460.80
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.2
[pip3] numpydoc==1.1.0
[pip3] torch==1.8.1
[pip3] torch-cluster==1.5.8
[pip3] torch-geometric==1.6.3
[pip3] torch-scatter==2.0.5
[pip3] torch-sparse==0.6.8
[pip3] torch-spline-conv==1.2.0
[pip3] torchaudio==0.8.0a0+e4e171a
[pip3] torchvision==0.9.1
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               11.1.1               h6406543_8    conda-forge
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2020.4             h726a3e6_304    conda-forge
[conda] mkl-service               2.3.0            py38h1e0a361_2    conda-forge
[conda] mkl_fft                   1.3.0            py38h5c078b8_1    conda-forge
[conda] mkl_random                1.2.0            py38hc5bc63f_1    conda-forge
[conda] numpy                     1.19.2           py38h54aff64_0  
[conda] numpy-base                1.19.2           py38hfa32c7d_0  
[conda] numpydoc                  1.1.0                      py_1    conda-forge
[conda] pytorch                   1.8.1           py3.8_cuda11.1_cudnn8.0.5_0    pytorch
[conda] torch-cluster             1.5.8                    pypi_0    pypi
[conda] torch-geometric           1.6.3                    pypi_0    pypi
[conda] torch-scatter             2.0.5                    pypi_0    pypi
[conda] torch-sparse              0.6.8                    pypi_0    pypi
[conda] torch-spline-conv         1.2.0                    pypi_0    pypi
[conda] torchaudio                0.8.1                      py38    pytorch
[conda] torchvision               0.9.1                py38_cu111    pytorch

**Comment by gchanan** [comment]
I didn't look closely into this, but it's probably something like https://github.com/pytorch/pytorch/issues/58128 -- it's not a bug.

**Comment by Xin-yang-Liu** [comment]
> I didn't look closely into this, but it's probably something like #58128 -- it's not a bug.

Thanks for your reply, I have change the title to numerical-reproducibility issue but I don't think it is the same as [#58128](https://github.com/pytorch/pytorch/issues/58128). This issue takes place not only when batch size varies but also when the matrix is of a certain type of sparse matrix.

**Comment by eqy** [comment]
> > I didn't look closely into this, but it's probably something like #58128 -- it's not a bug.
> 
> Thanks for your reply, I have changed the title to numerical-reproducibility issue but I don't think it is the same as [#58128](https://github.com/pytorch/pytorch/issues/58128). This issue takes place not only when batch size varies but also when the matrix is of a certain type of sparse matrix.

Unfortunately numerical precision issues will also be data-dependent, so without exhaustively testing both shapes and sparsity in isolation it's difficult to pin down the cause to precisely one issue. Does setting `NVIDIA_TF32_OVERRIDE=0` or switching to double precision fix the issue?

**Comment by Xin-yang-Liu** [comment]
> Unfortunately numerical precision issues will also be data-dependent, so without exhaustively testing both shapes and sparsity in isolation it's difficult to pin down the cause to precisely one issue. Does setting `NVIDIA_TF32_OVERRIDE=0` or switching to double precision fix the issue?

Changing data format/precision does not work. And this issue takes place regardless of whether it is running on the CPU or GPU. You could play with the sample code I provided. 

According to my test:

> if you change: 
> * the batch size of ```x``` back to ```1```(```x=torch.rand(1,64,1)```)
> * the ```periodic``` parameter in ```Build_CoefficientMatrix``` to ```False```
> * the ```c``` parameter in ```Build_CoefficientMatrix``` to a list with length no greater than 3
> 
> the two results match each other.

**Comment by ngimel** [comment]
This is due to tf32, please try disabling it https://pytorch.org/docs/master/backends.html#torch.backends.cuda.matmul.allow_tf32. I cannot repro on a V100 system.

**Comment by Xin-yang-Liu** [comment]
> This is due to tf32, please try disabling it https://pytorch.org/docs/master/backends.html#torch.backends.cuda.matmul.allow_tf32. I cannot repro on a V100 system. Also, what @gchanan said, you should not expect bitwise accurate results for batch matrix multiply and a regular matrix multiply, neither on cpu nor on cuda.

I don't think this issue is due to TF32 or any other CUDA related factors. This issue will raise both on GPU and CPU and regardless of dtype. 

According to @albanD in https://github.com/pytorch/pytorch/issues/58128, such issue is caused by 

>  low level compute libraries that we use might choose a different algorithm.

But bit-wise accuracy is really important to my current Reinforcement learning task where torch.matmul is iteratively used because one episode consists of hundreds of steps. Such inaccuracy will  accumulate and the results are unacceptable.

I hope you could add an option to force the low level libs to choose a deterministic algorithm even if this algorithm is not the most efficient one. Such option is really useful in RL tasks.

**Comment by ngimel** [comment]
They are using deterministic algorithms, just when you are computing batched matmul or single matmul they deterministically produce slightly different results. There's literally no way to force low level libs to produce same results in this case. If you want, you can implement matmul that would loop over batch elements yourself, but it will be catastrophic for performance.

**Comment by Xin-yang-Liu** [comment]
I have no idea how the low level libs work but in numpy, batched matmul does not have this issue. Is it possible to mimic np.matmul in pytorch?

**Comment by albanD** [comment]
@ngimel I think the other problem here as well is that since the input has different number of dimensions, it goes to `mm` in one case and `bmm`.
And if the other input was 2d, it would even use `mv`.

**Comment by albanD** [comment]
> I have no idea how the low level libs work but in numpy, batched matmul does not have this issue.

It depends on the shapes, I think numpy has the same behavior of switching between implementation as well.

```python
import torch
import numpy as np

a = torch.rand(20, 20)
b = torch.rand(20, 10)

batch_out = torch.matmul(a, b)[:, 0]
single_out = torch.matmul(a, b[:, 0])
print("Torch diff: ", (batch_out - single_out).abs().max())

a = a.numpy()
b = b.numpy()
np_batch_out = np.matmul(a, b)[:, 0]
np_single_out = np.matmul(a, b[:, 0])
print("Numpy diff: ", np.max(np.abs(np_batch_out - np_single_out)))

```

On colab, this returns:
```
Torch diff:  tensor(9.5367e-07)
Numpy diff:  9.536743e-07
```

**Comment by Xin-yang-Liu** [comment]
> It depends on the shapes, I think numpy has the same behavior of switching between implementation as well.
> 
> ```python
> import torch
> import numpy as np
> 
> a = torch.rand(20, 20)
> b = torch.rand(20, 10)
> 
> batch_out = torch.matmul(a, b)[:, 0]
> single_out = torch.matmul(a, b[:, 0])
> print("Torch diff: ", (batch_out - single_out).abs().max())
> 
> a = a.numpy()
> b = b.numpy()
> np_batch_out = np.matmul(a, b)[:, 0]
> np_single_out = np.matmul(a, b[:, 0])
> print("Numpy diff: ", np.max(np.abs(np_batch_out - np_single_out)))
> ```
> 
> On colab, this returns:
> 
> ```
> Torch diff:  tensor(9.5367e-07)
> Numpy diff:  9.536743e-07
> ```

I think we are not talking about the same thing. In your code, `batch_out` is a result of of matrix product while `single_out` is a matrix-vector product. Such difference is somewhat explainable.

What I think unacceptable is the difference in "batched matrix multiplication". 
```Python
import torch
import numpy as np

a = torch.rand(20, 20)
b = torch.rand(2, 20, 10)

batch_out = torch.matmul(a, b)[0, :, 0]
single_out = torch.matmul(a, b[0])[:,0]
print("Torch diff: ", (batch_out - single_out.squeeze()).abs().max())

a = a.numpy()
b = b.numpy()
np_batch_out = np.matmul(a, b)[0, :, 0]
np_single_out = np.matmul(a, b[0])[:,0]
print("Numpy diff: ", np.max(np.abs(np_batch_out - np.squeeze(np_single_out))))
```

On colab, this returns:
 ```
Torch diff:  tensor(4.7684e-07)
Numpy diff:  0.0
```
In this situation, what I expect is calculating `a*b[i]` in parallel. The difference here is unacceptable.

**Comment by albanD** [comment]
I used that because that's what the original code sample that you shared is doing.

For this particular case, you hit a special optimization that is done by transposing the matrices before computation. Note that you can trigger the same thing on numpy by providing an output matrix whose underlying storage is transpoed:
```python
import torch
import numpy as np

a = torch.rand(20, 20)
b = torch.rand(2, 20, 10)
output = torch.rand(2, 10, 20).transpose(-1, -2)

batch_out = torch.matmul(a, b)[0]
single_out = torch.matmul(a, b[0])
single_out2 = torch.matmul(b[0].T, a.T).T
print("Torch diff batch vs single: ", (batch_out - single_out).abs().max())
print("Torch diff single vs single transpose: ", (single_out - single_out2).abs().max())
print("Torch diff batch vs single transpose: ", (batch_out - single_out2).abs().max())

a = a.numpy()
b = b.numpy()
output = output.numpy()
np_batch_out = np.matmul(a, b, out=output)[0]
np_single_out = np.matmul(a, b[0])

print("Numpy diff: ", np.max(np.abs(np_batch_out - np_single_out)))
```
returns
```

import torch
import numpy as np

a = torch.rand(20, 20)
b = torch.rand(2, 20, 10)
output = torch.rand(2, 10, 20).transpose(-1, -2)

batch_out = torch.matmul(a, b)[0]
single_out = torch.matmul(a, b[0])
single_out2 = torch.matmul(b[0].T, a.T).T

Torch diff batch vs single:  tensor(9.5367e-07)
Torch diff single vs single transpose:  tensor(9.5367e-07)
Torch diff batch vs single transpose:  tensor(0.)
Numpy diff:  1.4305115e-06
```


But the main thing beyond all the details and different shape/tranposition. Natalia's original point remain: you cannot rely on the fact that slice/different number of dims of inputs will give you bit-wise exact results. Only functions with the exact same input (and with deterministic flag enabled) have a change to give the same result as long as you don't change any of the hardware or dependency version.

**Comment by Xin-yang-Liu** [comment]
> For this particular case, you hit a special optimization that is done by transposing the matrices before computation.
Is it possible to disable such optimization by default in PyTorch?

**Comment by albanD** [comment]
No. And while this particular one is actually in pytorch codebase, the underlying BLAS library might be doing similar tricks. On which we have no control.

**Comment by stadlerb** [comment]
The following example computes the maximum norm and Eucliean norm of the difference between batch size 1 and 2 for a random vector (duplicated in batch size 2) and a random matrix, with varying dimensions. It plots them for 32 to 8192 dimensions, in steps of 32:

```python
from statistics import mean

import matplotlib.pyplot as plt
import torch


def repro_diff(dim, seed=1):
    # Start with same seed for different dimensions to make results more comparable for different dimensions
    torch.manual_seed(seed)
    tensor_1 = torch.rand(1, dim)
    tensor_2 = tensor_1.expand(2, dim)
    matrix = torch.rand(dim, dim)
    result_unbatched = tensor_1.matmul(matrix)[0, :]
    result_batched = tensor_2.matmul(matrix)[0, :]
    return result_unbatched - result_batched


step_size = 32
num = 256
res = []

for i in range(num):
    dim = step_size * (i + 1)
    # Average over 3 different random seeds to reduce noise
    diffs = [repro_diff(dim, seed) for seed in [1, 2, 3]]
    mean_norm_diff = mean(diff.norm().item() for diff in diffs)
    mean_max_diff = mean(diff.abs().max().item() for diff in diffs)
    res.append((dim, mean_norm_diff, mean_max_diff))

dims, diff_eucl_means, diff_max_means = zip(*res)

fig, ax = plt.subplots()
ax.scatter(dims, diff_eucl_means, label='Mean euclidean norm of difference', alpha=0.5, s=10.0)
ax.scatter(dims, diff_max_means, label='Mean maximum norm of difference', alpha=0.5, s=10.0)
plt.xlabel('# Dimensions')
plt.yscale('log')
ax.legend()
ax.grid(True)
plt.show()
```
![matmul_inaccuracies](https://user-images.githubusercontent.com/2452384/125599055-4bfdd184-1283-4e21-a3e1-db64f8fe1e1d.png)

The differences that are significantly larger than 1e-6.  There is a jump downward at 512, which is reversed at 2048, maybe to different algorithms being selected internally in the range between.
